core/src/main/java/io/grpc/ChannelImpl.java

```diff
@@ -195,12 +204,13 @@ public ChannelImpl shutdown() {
    * Initiates a forceful shutdown in which preexisting and new calls are cancelled. Although
    * forceful, the shutdown process is still not instantaneous; {@link #isTerminated()} will likely
    * return {@code false} immediately after this method returns.
-   *
-   * <p>NOT YET IMPLEMENTED. This method currently behaves identically to shutdown().
    */
-  // TODO(ejona86): cancel preexisting calls.
   public ChannelImpl shutdownNow() {
     shutdown();
+
+    for (ClientCallImpl<?, ?> call : activeCalls) {
+      call.getStream().cancel(Status.INTERNAL.withDescription("Shutting down"));
```

Comment:
Before the awaitTerminated()+daemon stuff came up, I was completely expecting to add a shutdownNow() method to transports. That 1) doesn't require _any_ extra synchronization for transports, because transports already have to maintain a set of streams, 2) doesn't require _any_ extra bookkeeping, because transports already have to maintain a set of streams, 3) makes it more obvious atomic shutdown() is required from the transport, 4) can be considerably more efficient.

Because of the awaitTerminating stuff, 1 and 2 go away, and we don't really care about 4 too much. We may simply choose not to implement shutdownNow() now.

> If you feel there is too high of a cost to synchronizing Call, it will push the synchronization on the the application

We aren't synchronizing Call because of the belief that even if it is synchronized, the application still has to do synchronization. It also isn't thread-safe because that means that all interceptors must return thread-safe objects.

Effectively, you get _every_ layer doing synchronization, which seems a road better avoided.

> just say that shutdownNow is not thread safe. In this case, we can just say that shutdown is the preferred method.

That's basically the same as killing shutdownNow, because it really doesn't guarantee anything. But shutdownNow is necessary and needed. I don't think that is appropriate, especially since we do have the power to make it thread-safe without any heavy-weight solutions.